File: BluePay iOS SDK-ObjC/BluePay iOS SDK/BluePay.m
//
//  BluePay_Helper.m
//  BluePay Apple Pay SDK
//
//  Created by Justin Slingerland on 3/16/15.
//  Copyright (c) 2015 BluePay Processing, Inc. All rights reserved.
//

#import "BluePay.h"

@implementation BluePay

NSMutableDictionary *bluepaySetup;
NSString *accountID = @"Merchant's Account ID Here"; // 12 digit Account ID
NSString *secretKey = @"Merchant's Secret Key Here"; // 32 digit Secret Key
NSString *transMode = @"TEST"; // TEST or LIVE mode
NSString *transType = @"SALE"; // SALE or AUTH; defaults to SALE unless explicitly specified

- (id)initSetup:(NSString*)transactionType
{
    self = [super init];
    bluepaySetup = [[NSMutableDictionary alloc] init];
    [bluepaySetup setObject:accountID forKey:@"AccountID"];
    [bluepaySetup setObject:secretKey forKey:@"SecretKey"];
    [bluepaySetup setObject:transMode forKey:@"TransMode"];
    if (transactionType != nil && ![transactionType  isEqual: @""]) {
        [bluepaySetup setObject:transactionType forKey:@"TransType"];
    } else {
        [bluepaySetup setObject:transType forKey:@"TransType"];
    }
    return self;
}

- (NSMutableDictionary*) getBluePaySetup
{
    return bluepaySetup;
}

+ (NSString *)calcTPS:(NSString *)secretKey accID:(NSString *)accountID transType:(NSString *)transactionType transAmount:(NSString *)amount fullName: (NSString *) name paymentAcct:(NSString *)paymentAccount
{
    // Calculates the TAMPER_PROOF_SEAL needed for each transaction for the bp20post API
    NSMutableString *tps = [NSMutableString string];
    [tps appendString:secretKey];
    [tps appendString:accountID];
    [tps appendString:transactionType];
    [tps appendString:amount];
    [tps appendString:name];
    [tps appendString:paymentAccount];
    NSString *md5 = [tps MD5];
    return md5;
}

+ (NSDictionary *)getCustomerInformation:(PKPayment *) payment
{

    //CFTypeRef addressProperty = ABRecordCopyValue((__bridge ABRecordRef)billingInfo, kABPersonAddressProperty);
    //NSLog(@"%@",CNPostalAddressPostalCodeKey);
    //NSMutableDictionary *customerInformation = (__bridge NSMutableDictionary *)CFArrayGetValueAtIndex((CFArrayRef)ABMultiValueCopyArrayOfAllValues(addressProperty), 0);
    // Grab the first and last name values from the billing address
    //NSString *firstName = (__bridge_transfer NSString *)ABRecordCopyValue((__bridge ABRecordRef)(billingInfo), kABPersonFirstNameProperty);
    //NSString *firstName = billingInfo.
    //NSString *lastName = (__bridge_transfer NSString *)ABRecordCopyValue((__bridge ABRecordRef)(billingInfo), kABPersonFirstNameProperty);
    NSString *firstName = payment.billingContact.name.givenName;
    NSString *lastName = payment.billingContact.name.familyName;
    
    // Grab the phone and email values from the shipping address
    //ABMultiValueRef phones = ABRecordCopyValue((__bridge ABRecordRef)(shippingInfo), kABPersonPhoneProperty);
    //NSString *phone = (__bridge NSString *)(ABMultiValueCopyValueAtIndex(phones, 0));
    //NSString phone = billingInfo.
    //ABMultiValueRef emails = ABRecordCopyValue((__bridge ABRecordRef)(shippingInfo), kABPersonEmailProperty);
    //NSString *email = (__bridge NSString *)(ABMultiValueCopyValueAtIndex(emails, 0));
    NSString *phone = payment.shippingContact.phoneNumber.stringValue;
    NSString *email = payment.shippingContact.emailAddress;
    
    // Add customer first name, last name, phone, and email to the customerInformation NSDict
    NSMutableDictionary *customerInformation = [[NSMutableDictionary alloc] init];
    [customerInformation setValue:firstName forKey:@"FirstName"];
    [customerInformation setValue:lastName forKey:@"LastName"];
    [customerInformation setValue:phone forKey:@"Phone"];
    [customerInformation setValue:email forKey:@"Email"];
    return customerInformation;
}

@end

@implementation NSString (BluePayHelper)
- (NSString *)MD5
{
    const char *cstr = [self UTF8String];
    unsigned char result[16];
    CC_MD5(cstr, (CC_LONG)strlen(cstr), result);
    
    return [NSString stringWithFormat:
            @"%02X%02X%02X%02X%02X%02X%02X%02X%02X%02X%02X%02X%02X%02X%02X%02X",
            result[0], result[1], result[2], result[3],
            result[4], result[5], result[6], result[7],
            result[8], result[9], result[10], result[11],
            result[12], result[13], result[14], result[15]
            ];  
}
@end

@implementation NSString (URLEncoding)
- (NSString *)stringByAddingPercentEncodingForRFC3986 {
    NSString *unreserved = @"-._~/";
    NSMutableCharacterSet *allowed = [NSMutableCharacterSet
                                      alphanumericCharacterSet];
    [allowed addCharactersInString:unreserved];
    return [self
            stringByAddingPercentEncodingWithAllowedCharacters:
            allowed];
}
@end

File: BluePay iOS SDK-Swift/BluePay iOS SDK-Swift/BluePay.swift
import Foundation
import CommonCrypto
import AddressBook
import PassKit
class BluePay {
    var bluepaySetup = [String: String]()
    var AccountID: String = "100013391447" // 12 digit Account ID
    var SecretKey: String = "5YRFNRBCZN/6Y4OPZNWPYDRNAVX7BMMD" // 32 digit Secret Key
    var TransMode: String = "TEST" // TEST or LIVE mode
    var TransType: String = "SALE" // SALE or AUTH; defaults to SALE unless explicitly specified

    init(transactionType: String? = nil) {
        bluepaySetup = ["AccountID" : self.AccountID,
                        "SecretKey" : self.SecretKey,
                        "TransMode" : self.TransMode,
                        "TransType" : self.TransType]
        if transactionType != nil {
            bluepaySetup["TransType"] = transactionType
        }
    }

    func getBluePaySetup() -> NSDictionary {
        return bluepaySetup
    }

    class func calcTPS(secretKey: String, accID accountID: String, transType transactionType: String, transAmount amount: String, fullName name: String, paymentAcct paymentAccount: String) -> String {
            // Calculates the TAMPER_PROOF_SEAL needed for each transaction for the bp20post API
        var tps = ""
        tps += secretKey
        tps += accountID
        tps += transactionType
        tps += amount
        tps += name
        tps += paymentAccount
        let md5 = MD5(tps)
        return md5
    }

    class func getCustomerInformation(payment: PKPayment) -> [String : String] {
        // Grab the first and last name plus address values from the billing address
        let firstName: String = payment.billingContact!.name!.givenName!
        let lastName: String = payment.billingContact!.name!.familyName!
        let addr1: String = payment.billingContact!.postalAddress!.street
        let city: String = payment.billingContact!.postalAddress!.city
        let state: String = payment.billingContact!.postalAddress!.state
        let zip: String = payment.billingContact!.postalAddress!.postalCode
        let country: String = payment.billingContact!.postalAddress!.country
        // Grab the phone and email values from the shipping address
        let phone: String = payment.shippingContact!.phoneNumber!.stringValue
        let email: String = payment.shippingContact!.emailAddress!
        // Add customer first name, last name, phone, and email to the customerInformation NSDict
        var customerInformation: [String : String] = [String : String]()
        customerInformation["FirstName"] = firstName
        customerInformation["LastName"] = lastName
        customerInformation["Street"] = addr1
        customerInformation["City"] = city
        customerInformation["State"] = state
        customerInformation["ZIP"] = zip
        customerInformation["Country"] = country
        customerInformation["Phone"] = phone
        customerInformation["Email"] = email
        return customerInformation
    }
}

    func MD5(tps: String) -> String {
        let context = UnsafeMutablePointer<CC_MD5_CTX>.alloc(1)
        var digest = Array<UInt8>(count:Int(CC_MD5_DIGEST_LENGTH), repeatedValue:0)
        CC_MD5_Init(context)
        CC_MD5_Update(context, tps,
                      CC_LONG(tps.lengthOfBytesUsingEncoding(NSUTF8StringEncoding)))
        CC_MD5_Final(&digest, context)
        context.dealloc(1)
        var hexString = ""
        for byte in digest {
            hexString += String(format:"%02x", byte)
        }
        return hexString
    }

extension NSString {
    func stringByAddingPercentEncodingForRFC3986() -> String {
        let unreserved: String = "-._~/"
        let allowed: NSMutableCharacterSet = NSMutableCharacterSet.alphanumericCharacterSet()
        allowed.addCharactersInString(unreserved)
        return self.stringByAddingPercentEncodingWithAllowedCharacters(allowed)!
    }
}

extension NSData {
    
    var hexString : String {
        let buf = UnsafePointer<UInt8>(bytes)
        let charA = UInt8(UnicodeScalar("a").value)
        let char0 = UInt8(UnicodeScalar("0").value)
        
        func itoh(i: UInt8) -> UInt8 {
            return (i > 9) ? (charA + i - 10) : (char0 + i)
        }
        
        let p = UnsafeMutablePointer<UInt8>.alloc(length * 2)
        
        for i in 0..<length {
            p[i*2] = itoh((buf[i] >> 4) & 0xF)
            p[i*2+1] = itoh(buf[i] & 0xF)
        }
        
        return NSString(bytesNoCopy: p, length: length*2, encoding: NSUTF8StringEncoding, freeWhenDone: true)! as String
    }
}